### PR TITLE
New interface for binlog primary callbacks

### PR DESCRIPTION
**Pull request by fulghum**
First pass on connecting the GMS layer with the Dolt layer for handling callbacks when the SQL server is acting in binlog primary mode, through the new `BinlogPrimaryController` interface. This new interface pretty closely mirrors the existing callback interface for replica callbacks, the `BinlogReplicaController` interface.

Related to https://github.com/dolthub/dolt/issues/7512

### CODE REVIEW

File: enginetest/queries/queries.go [hunk 1]
```diff
@@ -9285,12 +9285,6 @@ from typestable`,
 			{"1.5000"},
 		},
 	},
-	{
```
Comment (zachmu) [inline enginetest/queries/queries.go:9288]: Where did this test go?

File: sql/binlogreplication/binlog_replication.go [hunk 2]
```diff
@@ -62,6 +64,31 @@ type BinlogReplicaController interface {
 	ResetReplica(ctx *sql.Context, resetAll bool) error
 }
 
+// BinlogPrimaryController allows GMS to coordinate with a binlog primary server. Providers built on go-mysql-server may
```
Comment (zachmu) [inline sql/binlogreplication/binlog_replication.go:67]: This first sentence makes it sound like this interface is used to communicate with a MySQL primary, when it's actually used when this server wants to be the primary. Maybe reword a bit for clarity here?

File: sql/binlogreplication/binlog_replication.go [hunk 3]
```diff
@@ -62,6 +64,31 @@ type BinlogReplicaController interface {
...
+	// primary server. |replicaHost| and |replicaPort| specify where the replica can be accessed, and are returned
+	// from the SHOW REPLICAS statement.
+	RegisterReplica(ctx *sql.Context, c *mysql.Conn, replicaHost string, replicaPort uint16) error
+
+	// BinlogDumpGtid tells the binlog primary controller to start streaming binlog events to the replica over
+	// connection |c|. |gtidSet| specifies the point at which to start replication, or if it is nil, then it indicates
+	// the complete history of all transactions should be sent over the connection.
+	BinlogDumpGtid(ctx *sql.Context, c *mysql.Conn, gtidSet mysql.GTIDSet) error
+
+	// ListReplicas is called when the SHOW REPLICAS statement is executed.
```
Comment (zachmu) [inline sql/binlogreplication/binlog_replication.go:82]: Any notes on the behavior here? Is this just a callback, since there's no output?

Same comment on the method below this

**Review by zachmu — APPROVED** [review]
LGTM, just a couple comments

File: enginetest/queries/queries.go (hunk 1 above)
Comment (fulghum) [reply enginetest/queries/queries.go:9288]: `server_uuid` was getting initialized to `uuid.New()` each time the server started. This ensured there was always a server UUID, but it wasn't useful, since it changed every time the server restarted. As a binlog replica, it didn't make a difference, but as a binlog primary, it does since the replica uses the primary server's UUID as part of the GTID set identifier. 

`server_uuid` now defaults to `nil` and users will need to set or persist it before they can use replication. I think that's okay for now, but I'd like to move closer to MySQL's behavior where `server_uuid` is initialized the first time a server is started and automatically persisted. I think we shouldn't persist it in the standard `~/.dolt` data director for persisted system vars though, since then it's the `server_uuid` for any server that host starts. This is on my list to come back to, but I think simply defaulting to `nil` for now and having users set it explicitly is alright in the meantime.

File: sql/binlogreplication/binlog_replication.go (hunk 3 above)
Comment (fulghum) [reply sql/binlogreplication/binlog_replication.go:82]: I'll add some more explanation. I mostly wanted to get those stubbed out in the interface since I know GMS will need to call into the integrator's code for those, but they will still need to change to return some data to GMS.
